pkg/kubelet/cpumanager/cpu_manager.go

```diff
@@ -0,0 +1,363 @@
+
+	IsUnderCPUPressure() bool
+
+	lifecycle.PodAdmitHandler
+}
+
+type manager struct {
+	sync.Mutex
+	policy Policy
+	state  state.State
+
+	// containerRuntime is the container runtime service interface needed
+	// to make UpdateContainerResources() calls against the containers.
+	containerRuntime runtimeService
+
+	// kletGetter provides a method for listing all the pods on the node
+	// so all the containers can be updated in the reconciliation loop.
+	kletGetter kletGetter
+
+	// podStatusProvider provides a method for obtaining pod statuses
+	// and the containerID of their containers
+	podStatusProvider status.PodStatusProvider
+
+	// getPodsFunc gets list of active pods to consider for eviction when
+	// under pressure.
+	getPodsFunc eviction.ActivePodsFunc
+
+	// killPodFunc kills a pod selected for eviction when under pressure.
+	killPodFunc eviction.KillPodFunc
+
+	// recorder records eviction events.
+	recorder record.EventRecorder
+}
+
+var _ Manager = &manager{}
+
+// NewManager creates new cpu manager based on provided policy
+func NewManager(
+	cpuPolicyName string,
+	cr runtimeService,
+	kletGetter kletGetter,
+	statusProvider status.PodStatusProvider,
+	capacityProvider eviction.CapacityProvider,
+	getPodsFunc eviction.ActivePodsFunc,
+	killPodFunc eviction.KillPodFunc,
+	recorder record.EventRecorder,
+) (Manager, lifecycle.PodAdmitHandler, error) {
+	var policy Policy
+
+	switch policyName(cpuPolicyName) {
+	case PolicyNone:
+		policy = NewNonePolicy()
+	case PolicyStatic:
+		machineInfo, err := kletGetter.GetCachedMachineInfo()
+		if err != nil {
+			return nil, nil, err
+		}
+		topo, err := topology.Discover(machineInfo)
+		if err != nil {
+			return nil, nil, err
+		}
+		glog.Infof("[cpumanager] detected CPU topology: %v", topo)
+
+		resources := capacityProvider.GetNodeAllocatableReservation()
+		cpuResource, ok := resources[v1.ResourceCPU]
+		if ok {
+			reservedCores := int(cpuResource.Value())
+			glog.Infof("[cpumanager] reserving %v cores due to kube/system reserved", reservedCores)
+			topo.NumReservedCores = reservedCores
+		}
+		policy = NewStaticPolicy(topo)
+	default:
+		glog.Warningf("[cpumanager] Unknown policy (\"%s\"), falling back to \"%s\" policy (\"%s\")", cpuPolicyName, PolicyNone)
+		policy = NewNonePolicy()
+	}
+
+	manager := &manager{
+		policy:            policy,
+		state:             state.NewMemoryState(),
+		containerRuntime:  cr,
+		kletGetter:        kletGetter,
+		podStatusProvider: statusProvider,
+		getPodsFunc:       getPodsFunc,
+		killPodFunc:       killPodFunc,
+		recorder:          recorder,
+	}
+	return manager, manager, nil
+}
+
+func (m *manager) Start() {
+	glog.Infof("[cpumanger] starting with %s policy", m.policy.Name())
+	m.policy.Start(m.state)
+	if m.policy.Name() == string(PolicyNone) {
+		return
+	}
+	go wait.Until(func() { m.reconcileState() }, time.Second, wait.NeverStop)
+}
+
+func (m *manager) RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error {
```

Comment:
this flow was not obvious to me, but i have a better understanding of the intent.

my high-level understanding:

kubelet syncPod
- kubelet.go knows about containerManager only
- it asks a containerManager for a "PodContainerManager"
- it asks pod container manager to verify a pod cgroup exists (if needed)
- kubelet then calls containerruntime.SyncPod(...)
- container runtime sync pod creates container
- container runtime sync pod then registers with cpu manager
- cpu manager then calls back to container runtime and updates container resources with new cpu set assignment
- container runtime then starts container

let me think on a better flow on this over night, but i understand that this may be non-obvious.